ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java

```diff
@@ -298,21 +347,39 @@ void onErrorResponse(ReplicaId replicaId) {
   }
 
   /**
-   * Handle the body of the response: Deserialize and set the {@link BlobInfo} to return.
+   * Handle the body of the response: Deserialize and set the {@link BlobInfo} to return if no decryption is required.
+   * If decryption is required, submit a jon for decryption.
    * @param payload the body of the response.
    * @param messageMetadata the {@link MessageMetadata} associated with the message.
    * @throws IOException if there is an IOException while deserializing the body.
    * @throws MessageFormatException if there is a MessageFormatException while deserializing the body.
    */
   private void handleBody(InputStream payload, MessageMetadata messageMetadata)
       throws IOException, MessageFormatException {
-    ByteBuffer encryptionKey = messageMetadata == null ? null : messageMetadata.getEncryptionKey();
-    // @todo use the encryption key for decryption.
-    if (operationResult == null) {
-      BlobInfo blobInfo = new BlobInfo(MessageFormatRecord.deserializeBlobProperties(payload),
-          MessageFormatRecord.deserializeUserMetadata(payload).array());
-      getOptions().ageAtAccessTracker.trackAgeAtAccess(blobInfo.getBlobProperties().getCreationTimeInMs());
-      operationResult = new GetBlobResultInternal(new GetBlobResult(blobInfo, null), null);
+    if (!successfullyDeserialized) {
+      ByteBuffer encryptionKey = messageMetadata == null ? null : messageMetadata.getEncryptionKey();
+      serverBlobProperties = MessageFormatRecord.deserializeBlobProperties(payload);
+      serverUserMetadata = MessageFormatRecord.deserializeUserMetadata(payload);
+      successfullyDeserialized = true;
+      if (!serverBlobProperties.isEncrypted()) {
+        // if blob is not encrypted, move the state to Complete
+        state = State.Complete;
+        if (operationResult == null) {
+          operationResult = new GetBlobResultInternal(
+              new GetBlobResult(new BlobInfo(serverBlobProperties, serverUserMetadata.array()), null), null);
+        }
+      } else {
+        // submit decrypt job
+        state = State.Decrypting;
+        logger.trace("Moving the state to {} and submitting decrypt job for {}", State.Decrypting, blobId);
+        cryptoJobHandler.submitJob(
+            new DecryptJob(blobId, encryptionKey.duplicate(), null, serverUserMetadata, cryptoService, kms,
+                (DecryptJob.DecryptJobResult result, Exception exception) -> {
+                  logger.trace("Handling decrypt job callback results for {}", blobId);
+                  decryptCallbackResultInfo.get().setResultAndException(result, exception);
+                  routerCallback.onPollReady();
+                }));
+      }
     } else {
```

Comment:
this `else` and the comment beyond it don't apply anymore. This `else` is _specifically_ for the case when `operationResult` is not `null`